package-config.cmake

```diff
@@ -137,11 +137,49 @@ string(REPLACE ";" "\n  " PACKAGE_DEPENDENCIES_FIND_DEPENDENCY "${_PACKAGE_CONFI
   set(PACKAGE_EXTRA_MACROS "${PACKAGE_EXTRA_MACROS}
-if(COMMAND CHECK_MINIMAL_CXX_STANDARD)
-  CHECK_MINIMAL_CXX_STANDARD(${_MINIMAL_CXX_STANDARD})
-else()
-  MESSAGE(WARNING \"This dependency requires C++ >= ${_MINIMAL_CXX_STANDARD}\")
-endif()")
+set(_MINIMAL_CXX_STANDARD ${_MINIMAL_CXX_STANDARD})
+set(MINIMAL_CXX_STANDARD_ENFORCE ${MINIMAL_CXX_STANDARD_ENFORCE})
+
+# Get compiler default cxx standard, by printing \"__cplusplus\" (only once)
+if(NOT DEFINED _COMPILER_DEFAULT_CXX_STANDARD)
+  if(MSVC)
+    # See https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/
+    string(APPEND CMAKE_CXX_FLAGS \" /Zc:__cplusplus\")
+  endif()
+  
+  write_file($\{CMAKE_CURRENT_BINARY_DIR\}/tmp-cxx-standard.cpp \"#include <iostream>\\nint main(){std::cout << __cplusplus;return 0;}\")
+  try_run(_cxx_standard_run_status _cxx_standard_build_status
+    $\{CMAKE_CURRENT_BINARY_DIR\} $\{CMAKE_CURRENT_BINARY_DIR\}/tmp-cxx-standard.cpp
+    RUN_OUTPUT_VARIABLE _COMPILER_DEFAULT_CXX_STANDARD)
+  message(STATUS \"Default C++ standard: $\{_COMPILER_DEFAULT_CXX_STANDARD\}\")
+endif()
+
+# Check if a non-trivial minimum has been requested
+if(DEFINED _MINIMAL_CXX_STANDARD AND NOT _MINIMAL_CXX_STANDARD EQUAL 98)
+  # Check that the requested minimum is higher than the compiler default
+  # ref https://en.cppreference.com/w/cpp/preprocessor/replace#Predefined_macros for constants
+  if(_COMPILER_DEFAULT_CXX_STANDARD EQUAL 199711
+      OR (_COMPILER_DEFAULT_CXX_STANDARD EQUAL 201103 AND _MINIMAL_CXX_STANDARD GREATER 11)
+      OR (_COMPILER_DEFAULT_CXX_STANDARD EQUAL 201402 AND _MINIMAL_CXX_STANDARD GREATER 14)
+      OR (_COMPILER_DEFAULT_CXX_STANDARD EQUAL 201703 AND _MINIMAL_CXX_STANDARD GREATER 17)
+      OR (DEFINED CMAKE_CXX_STANDARD AND CMAKE_CXX_STANDARD LESS 98 AND _MINIMAL_CXX_STANDARD GREATER CMAKE_CXX_STANDARD)
+      OR (DEFINED CMAKE_CXX_STANDARD AND CMAKE_CXX_STANDARD EQUAL 98 AND _MINIMAL_CXX_STANDARD LESS CMAKE_CXX_STANDARD))
+    message(STATUS \"Incompatible C++ standard detected: Current standard is $\{CMAKE_CXX_STANDARD\} and minimal is $\{_MINIMAL_CXX_STANDARD\}\")
+    # Check that the requested minimum is higher than any pre-existing CMAKE_CXX_STANDARD
+    if(NOT CMAKE_CXX_STANDARD OR CMAKE_CXX_STANDARD EQUAL 98 OR CMAKE_CXX_STANDARD LESS _MINIMAL_CXX_STANDARD)
+      if(ENFORCE_MINIMAL_CXX_STANDARD OR MINIMAL_CXX_STANDARD_ENFORCE)
+        set(ORIGINAL_CMAKE_CXX_STANDARD ${CMAKE_CXX_STANDARD})
+        set(CMAKE_CXX_STANDARD ${_MINIMAL_CXX_STANDARD})
+        message(AUTHOR_WARNING \"CMAKE_CXX_STANDARD automatically upgraded to $\{CMAKE_CXX_STANDARD\} (from $\{ORIGINAL_CMAKE_CXX_STANDARD\})\")
+      else()
+        message(FATAL_ERROR \"CMAKE_CXX_STANDARD upgrade to >= $\{_MINIMAL_CXX_STANDARD\} required\")
+      endif()
+    endif()
+  else()
+    message(STATUS \"C++ standard sufficient: Minimal required $\{_MINIMAL_CXX_STANDARD\}, currently defined: $\{CMAKE_CXX_STANDARD\}\")
+  endif()
+endif()
+")
```

Comment:
This looks like a copy paste of the `CHECK_MINIMAL_CXX_STANDARD` macro, I suggest we avoid this by:
- putting the body of the macro in a separate file (i.e. everything after `cmake_parse_arguments(...)` and before `endmacro()` that is included in cxx-standard.cmake
- read the content of this file and append it to the `PACKAGE_EXTRA_MACROS` if needed

Note: after looking more carefully, the bit in the macro that sets `_MINIMAL_CXX_STANDARD` should be out of the included file too

See [here](https://github.com/gergondet/include-in-cmake-macro-example) for an example of what I mean

Reply:
I agree this should be cleaner. The logic is not a full copy paste and has been adapted (the original one copy-paste did not work). The two key differences are (as far as I can tell/remember since time has passed):

1. The block around `# Check if we need to upgrade the current minimum` has been removed
2. The logic inside the block of `# Check if a non-trivial minimum has been requested` has been adapted (some of the variables won't be available in the packaged version - some we'd like to remain variables and others we'd like to replaced - cf. use of backslashes)

I'll spend some time today to make this cleaner - however due to time constraints and requirements to check the downstream created files with different configurations this may not be completed.

Reply:
I found a workable solution and tested with several different configurations - thank you for the pointers and feedback. In the process I've also augmented the existing status/warning message output.